pybricks/common/pb_type_speaker.c

```diff
@@ -22,34 +23,90 @@
 #include <pybricks/util_mp/pb_obj_helper.h>
 #include <pybricks/util_pb/pb_error.h>
 
+#define MP_PI MICROPY_FLOAT_CONST(3.14159265358979323846)
+
 typedef struct {
     mp_obj_base_t base;
     bool intialized;
 } pb_type_Speaker_obj_t;
 
 STATIC pb_type_Speaker_obj_t pb_type_Speaker_singleton;
 
+STATIC uint8_t volume = 100; // volume in 0..100 range
+
+// The number to multiply the sample amplitude by, to amplify the signal by the defined volume.
+// The original sample amplitude must be in the -1..1 range.
+STATIC uint16_t sample_amplifier = INT16_MAX;
+
 STATIC uint16_t waveform_data[128];
 
+// The amount of radians each PCM sample represents to completely fill whe waveform buffer for 2*PI
+//STATIC mp_float_t pcm_sample_size_rad = MP_PI * 2 / MP_ARRAY_SIZE(waveform_data);
+
+STATIC mp_obj_t pb_type_Speaker_get_volume(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    return mp_obj_new_int(volume);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pb_type_Speaker_get_volume_obj, 1, pb_type_Speaker_get_volume);
```

Comment:
The pattern we have been using so far in Pybricks is to have a single method for getting and setting.

It would probably be best to start a pull request in the `pybricks-api` repo with the API and docs to work out those kinds of details first, then update the implementation here to match.

Reply:
I have the get/set methods ready on my local already. But not sure at all how to have a single method for getting and setting? It does not make sense to me yet, would you have an example?

Reply:
example: https://github.com/pybricks/pybricks-micropython/blob/36b53df4d05fe6eddff7d74bf2b79b9a15f9447f/pybricks/pupdevices/pb_type_pupdevices_remote.c#L264-L307

It has a variable number of arguments, 1 or 2. If it is 1, then return the value like the getter. If it is 2, then set the value and return None.

Reply:
Is this a common pattern in Python? Because this would be a massive anti-pattern in the JVM world 😆

Reply:
It is "old-school" python from before there were properties. It was once common, but not anymore. We are using is because we have very limited flash memory on some hubs and this makes the code as small as possible.

Reply:
This is interesting. Would you mind elaborating on:
1. why do we have smaller code when using a single method only? Asking because, with a single method, we have to have additional if/else logic which takes quite some space too.
2. how much space-saving does this bring? Asking because I'm curious how the stuff works. Most of the time, developers care about code readability. Here we optimise for memory/space which is a whole different and interesting world.
3. Could we utilise properties somehow to avoid having a single method dealing with two concerns?

Reply:
> I have the get/set methods ready on my local already. But not sure at all how to have a single method for getting and setting? It does not make sense to me yet, would you have an example?

For additional context, we have a few like these as well:

![image](https://user-images.githubusercontent.com/12326241/148553005-abebf21e-bef0-4895-ae26-6bb4e8f100cb.png)

We could make 6 individual getters and setters for this, or just have this one method. Besides space-savings on the firmware, the same applies to the end-user program. 

That said, it's worth adding that the direction of the project and space saving concerns evolved a bit over time. Some design patterns were established early on, and we do try to keep things somewhat consistent unless changing it is considerably better.

In this particular case, it's also worth considering how the speaker might work in the future. Would there be other speaker settings (waveform, ....? ), and if so, how would those be configured? Separately as done here or perhaps as `speaker.settings(volume, waveform)`?

This is why we usually work out API changes in `pybricks-api` first, and implement them later as David suggested above.

----

> How much space-saving does this bring? 

The build size is printed when you run the make command. We also track the [build size](https://stats.pybricks.com/primehub.html)

Reply:
API Change PR: https://github.com/pybricks/pybricks-api/pull/92

Reply:
The image size when using:
1. two get/set methods: `239485`
2. single merged get/set method: `239453`

I.e. saved 32 bytes.

Reply:
Refactored now to a single `volume` method.

Reply:
> I.e. saved 56 bytes.

This is exactly what I was going to guess. :grin: That is the overhead of declaring an extra MicroPython function object. The C functions are compareable is size, I'm sure, but it is the second `MP_DEFINE_CONST_FUN_OBJ_XXX()` that doubles the overhead.